pkg/ecdsa/dkg/states.go

```diff
@@ -107,35 +134,290 @@ func (skgs *symmetricKeyGenerationState) Receive(msg net.Message) error {
 }
 
 func (skgs *symmetricKeyGenerationState) Next() state.State {
-	return &finalizationState{
+	return &tssRoundOneState{
 		channel: skgs.channel,
-		member:  skgs.member.initializeFinalization(),
+		member:  skgs.member.initializeTssRoundOne(),
 	}
 }
 
 func (skgs *symmetricKeyGenerationState) MemberIndex() group.MemberIndex {
 	return skgs.member.id
 }
 
+// tssRoundOneState is the state during which members broadcast TSS
+// commitments and the Paillier public key.
+// `tssRoundOneMessage`s are valid in this state.
+type tssRoundOneState struct {
+	channel net.BroadcastChannel
+	member  *tssRoundOneMember
+
+	phaseMessages []*tssRoundOneMessage
+}
+
+func (tros *tssRoundOneState) DelayBlocks() uint64 {
+	return tssRoundOneStateDelayBlocks
+}
+
+func (tros *tssRoundOneState) ActiveBlocks() uint64 {
+	return tssRoundOneStateActiveBlocks
+}
+
+func (tros *tssRoundOneState) Initiate(ctx context.Context) error {
+	// The ctx instance passed as Initiate argument is scoped to the lifetime
+	// of the current state. However, the Initiate method is blocking and the
+	// ctx instance is cancelled properly only after Initiate returns. Because
+	// of that, we cannot use ctx as round timeout signal as Initiate would
+	// hang forever if something goes wrong. To avoid such a resource leak,
+	// we set a round timeout based on state block duration and an average
+	// block time. The exact duration doesn't need to be super-accurate because
+	// if the timeout is hit, the execution will fail anyway. We just want
+	// to give enough time for round computation and make sure the round
+	// terminates regardless of the result.
+	stateBlocks := tros.DelayBlocks() + tros.ActiveBlocks()
+	stateDuration := 15 * time.Duration(stateBlocks)
+	roundCtx, roundCtxCancel := context.WithTimeout(ctx, stateDuration)
+	defer roundCtxCancel()
```

Comment:
I am late to the party but there is something about `roundCtx` that makes me really uncomfortable. 

If the actual block time is less than 15 seconds and someone hangs on `member.tssRoundOne` or any other round execution, they will get out of sync with the rest of the group. They will keep executing the protocol with a delay as if nothing wrong happened. Sure, the other group members should reject messages from the operator that is late, but debugging problems might be a nightmare, especially since one operator can have multiple member seats, each one potentially running into their own out-of-sync problems.

The current design of the state machine assumes `Initiate` is a function that can execute the logic and quickly return. Any work that happens in the background - like retransmission - is done in a separate goroutine limited in time by the context and block waiter that cancels it.

There are two options I can think of we can use instead of creating another context and assuming the block has 15 seconds:
- Trigger time-consuming TSS rounds in a separate goroutine with execution time limited by the `ctx` passed to `Initiate`,
- Rework the state machine to not block on `Initiate` and to cancel the context when the block waiter reaches the desired height.

Reply:
> Actually, roundCtx is only for avoiding hanging forever in case something goes wrong and the round message cannot be produced. If the actual block time is less than 15 seconds, the entire state duration will be shorter than roundCtx and if the member does not produce the round message quickly, it will fail anyway. 

That's something I am not sure about. I don't think it will fail, it will rather keep executing the state machine, no?

That being said, I do not think this change breaks the protocol but I think we need to make it right (option 1 or 2) before we consider #3041 as done.

Reply:
> That's something I am not sure about. I don't think it will fail, it will rather keep executing the state machine, no?

Actually, due to TSS nature, it will go to the next state but fail really soon since it will miss required messages.

> That being said, I do not think this change breaks the protocol but I think we need to make it right (option 1 or 2) before we consider https://github.com/keep-network/keep-core/issues/3041 as done.

Yep, we will address it before closing 3041